File: javascript/extensions/datetime.js
exports.EsonDatetime = {
    name: "EsonDatetime",
    should_encode: value => value instanceof Date,
    encode: value => {
        // ESON processes datetime in micro-seconds
        return {
            timestamp: value.getTime() * 1000,
        }
    },
    decode: value => new Date(value.timestamp / 1000),
}

exports.EsonDate = {
    name: "EsonDate",
    // Javascript doesn't support separation of date and datetime
    should_encode: () => false,
    encode: () => null,
    decode: value => {
        return new Date(value.year, value.month - 1, value.day)
    },
}

File: python/eson/extensions/datetime.py
from datetime import datetime, timezone, timedelta
from eson.config import EsonExtension


class EsonDatetime(EsonExtension):
    def should_encode(self, value) -> bool:
        return isinstance(value, datetime)

    def encode(self, value):
        """Accepts a datetime object"""
        return {
            # Accurate to a micro second
            "timestamp": int(value.timestamp() * 1000000)
        }

    def decode(self, encoded_value):
        timestamp = encoded_value.get("timestamp")
        return datetime.fromtimestamp(timestamp / float(1000000))
